### PR TITLE
Lab02 - Implement bubble sort

### PR DESCRIPTION
**Pull request by delvein**
Fixes #42 

#### Changes proposed in this PR:

- This is to print the given numbers in sorted order using bubble sort algorithm.
- Also added a test.

### CODE REVIEW

File: 02_bubble_sort/bubblesort.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+				} else {
+					largerItems = append(largerItems, val)
+				}
+			}
+		}
+
+		bubble(smallerItems)
+		bubble(largerItems)
+
+		var merged []int = append(append(append([]int{}, smallerItems...), []int{items[pivotIndex]}...), largerItems...)
```
Comment (golangcibot) [inline 02_bubble_sort/bubblesort.go]: should omit type []int from declaration of var merged; it will be inferred from the right-hand side (from `golint`)

Comment (delvein) [reply 02_bubble_sort/bubblesort.go]: Fixed this issue.

File: 02_bubble_sort:selva/bubblesort.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package main
```
Comment (golangcibot) [inline 02_bubble_sort:selva/bubblesort.go]: 1-47 lines are duplicate of `02_bubble_sort_selva/bubblesort.go:1-47` (from `dupl`)

File: 02_bubble_sort_selva/bubblesort.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package main
```
Comment (golangcibot) [inline 02_bubble_sort_selva/bubblesort.go]: 1-47 lines are duplicate of `02_bubble_sort:selva/bubblesort.go:1-47` (from `dupl`)

File: 02_bubble_sort:selva/bubblesort.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(items []int) []int {
+
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: Unnecessary blank line

File: 02_bubble_sort_selva/bubblesort.go (hunk 3 above)
Comment (anzdaddy) [inline 02_bubble_sort_selva/bubblesort.go]: Remove this file.

File: 02_bubble_sort_selva/bubblesort_test.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package main
```
Comment (anzdaddy) [inline 02_bubble_sort_selva/bubblesort_test.go]: Remove this file.

File: 02_bubble_sort:selva/bubblesort.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		}
+
+		bubble(smallerItems)
+		bubble(largerItems)
+
+		var merged = append(append(append([]int{}, smallerItems...), []int{items[pivotIndex]}...), largerItems...)
+
+		for j := 0; j < len(items); j++ {
+			items[j] = merged[j]
+		}
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: Don't copy the result back into the input slice. Return it directly.

File: 02_bubble_sort:selva/bubblesort.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+	if len(items) > 1 {
+		pivotIndex := len(items) / 2
+		var smallerItems = []int{}
+		var largerItems = []int{}
+
+		for i := range items {
+			val := items[i]
+			if i != pivotIndex {
+				if val < items[pivotIndex] {
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: Cache the pivot element.

File: 02_bubble_sort:selva/bubblesort.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+func bubble(items []int) []int {
+
+	if len(items) > 1 {
+		pivotIndex := len(items) / 2
+		var smallerItems = []int{}
+		var largerItems = []int{}
+
+		for i := range items {
+			val := items[i]
+			if i != pivotIndex {
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: Avoid this test by simply allowing the pivot element to drop into `largerItems`.

File: 02_bubble_sort:selva/bubblesort.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(items []int) []int {
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: This a copying flavour of quicksort, not bubble sort. I've made recommendations below, anyway. I'd suggest you just rename it, since it's better than bubble sort and gets you the extra bonus points.

Double extra bonus points if you rework it to avoid copying, except at the top level, as required.

File: 02_bubble_sort:selva/bubblesort.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+				if val < items[pivotIndex] {
+					smallerItems = append(smallerItems, val)
+				} else {
+					largerItems = append(largerItems, val)
+				}
+			}
+		}
+
+		bubble(smallerItems)
+		bubble(largerItems)
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: To accommodate the copying change recommended below:
```go
smallerItems = bubble(smallerItems)
largerItems = bubble(largerItems)
```

File: 02_bubble_sort:selva/bubblesort.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+		var merged = append(append(append([]int{}, smallerItems...), []int{items[pivotIndex]}...), largerItems...)
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: `append([]int{}, smallerItems...)` is the same as `smallerItems`.

Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: `[]int{items[pivotIndex]}...` is the same as `items[pivotIndex]`. But it's not required anyway if you allow the pivot element to drop into `largerItems`, as recommended above.

File: 02_bubble_sort:selva/bubblesort.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+		for j := 0; j < len(items); j++ {
+			items[j] = merged[j]
+		}
+
+	}
+
+	return items
+
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: Unnecessary blank line

File: 02_bubble_sort:selva/bubblesort.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+		bubble(largerItems)
+
+		var merged = append(append(append([]int{}, smallerItems...), []int{items[pivotIndex]}...), largerItems...)
+
+		for j := 0; j < len(items); j++ {
+			items[j] = merged[j]
+		}
+
+	}
+
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: Unnecessary blank line

File: 02_bubble_sort:selva/bubblesort.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+		bubble(smallerItems)
+		bubble(largerItems)
+
+		var merged = append(append(append([]int{}, smallerItems...), []int{items[pivotIndex]}...), largerItems...)
+
+		for j := 0; j < len(items); j++ {
+			items[j] = merged[j]
+		}
+
```
Comment (anzdaddy) [inline 02_bubble_sort:selva/bubblesort.go]: Unnecessary blank line

**Review by anzdaddy — CHANGES_REQUESTED** [review]

File: 02_bubble/selva/bubblesort.go [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(items []int) []int {
+
```
Comment (juliaogris) [inline 02_bubble/selva/bubblesort.go:12]: Remove blank lines at beginning/end of block

File: 02_bubble/selva/bubblesort.go [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+	if len(items) > 1 {
+		pivotIndex := len(items) / 2
+		var smallerItems = []int{}
+		var largerItems = []int{}
+
+		for i := range items {
```
Comment (juliaogris) [inline 02_bubble/selva/bubblesort.go:18]: Something about this implementation seems a bit more verbose than necessary - look up reference implmentations please.

File: 02_bubble/selva/bubblesort_test.go [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestMainOutputpwd(t *testing.T) {
```
Comment (juliaogris) [inline 02_bubble/selva/bubblesort_test.go:11]: Write a function `TestBubble` that tests `bubble` directly.
Include corner cases: empty slice, 1-element slice, sorted slice, reverse-sorted slice, slice with duplicated elements.
Consider using [table driven tests](https://github.com/golang/go/wiki/TableDrivenTests)

**Review by juliaogris — COMMENTED** [review]
-   Click `Update branch` and locally `git pull origin lab2` before adding new commits.
-   Please edit your commits to follow [recommendations](https://chris.beams.io/posts/git-commit/)
    -   Use the imperative mood in the subject line
    -   Fold fixup commits into previous ones with `git rebase -i HASH`
I can give a hand with this if you want

File: 02_bubble_sort:selva/bubblesort.go (hunk 2 above)
Comment (delvein) [reply 02_bubble_sort:selva/bubblesort.go]: Removed the duplication.

File: 02_bubble_sort_selva/bubblesort.go (hunk 3 above)
Comment (delvein) [reply 02_bubble_sort_selva/bubblesort.go]: Removed the duplication.

Comment (delvein) [reply 02_bubble_sort_selva/bubblesort.go]: Removed.

**Comment by juliaogris** [comment]
Hi,
There has been  no activity in the last 20 days so I will close this PR.
Feel free to open it again if you have something to add.
Thanks!
Julia
